ophyd/controls/areadetector/detectors.py

```diff
@@ -39,9 +41,138 @@
            'URLDetector',
            ]
 
+def set_and_wait(signal, val):
+    """
+    Set a signal to a value and wait until it reads correctly.
+
+    There are cases where this would not work well, so it should be revisited.
+    """
+    signal.put(val)
+    while signal.get() != val:
+        ttime.sleep(0.1)
+        logger.info("Waiting for %s to be set...", signal.name)
+
 
 class DetectorBase(ADBase):
-    pass
+    "This base class handles the staging, unstaging, and triggering."
+
+    # OphydObj subscriptions
+    _SUB_ACQ_DONE = 'acq_done'
+    _SUB_TRIGGER_DONE = 'trigger_done'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # TODO Should we make these settings customizable in the init?
+        # If so, what API?
+
+        # settings
+        self._stage_sigs = {self.cam.acquire: 0,  # If acquiring, stop.
+                            self.cam.image_mode: 1  # 'Multiple' mode
+                           }
+
+        self._staged = False
+
+        self._acquisition_signal = self.cam.acquire  # for generality's sake
+        self._acquisition_signal.subscribe(self._acquire_changed)
+
+    def stage(self):
+        """
+        Setup the detector to be triggered.
+
+        This must be called once before any calls to 'trigger'.
+        Multiple calls (before unstaging) have no effect.
+        """ 
+        if self._staged:
+            return
+
+        # Read and stage current values, to be restored by unstage()
+        self._original_vals = {sig: sig.get() for sig in self._stage_sigs}
+
+        # Apply settings.
+        self._staged = True
+        for sig, val in self._stage_sigs.items():
+            set_and_wait(sig, val)
+
+        # Call stage() on child devices (including, notably, plugins).
+        for signal_name in self.signal_names:
+            signal = getattr(self, signal_name)
+            if hasattr(signal, 'stage'):
+                signal.stage()
+
+        self._trigger_counter = 0  # total acquisitions while staged
+
+    def unstage(self):
+        """
+        Restore the detector to 'standby'.
+
+        Multiple calls (without a new call to 'stage') have no effect.
+        """
+        if not self._staged:
+            return
+
+        # Restore original values.
+        for sig, val in self._original_vals.items():
+            set_and_wait(sig, val)
+
+        # Call unstage() on child devices (including, notably, plugins).
+        for signal_name in self.signal_names:
+            signal = getattr(self, signal_name)
+            if hasattr(signal, 'stage'):
+                signal.stage()
+
+        self._staged = False
+
+    def trigger(self):
```

Comment:
I think so.  The detector already has enough control to manage taking a dark frame (from the C layer) no point in leaving that on the table.

The advantage of making the detectors in charge of this is that you can use the same `Plan` with a wide range dark frame stratgies (as the DF logic is selected by setting state on the AD object).  The big selling point here is that BS becomes blissfully unaware of the ccd/IP vs detris/eiger/pilatus differences. 

Putting the logic in the AD opens up the possibility of it re-taking dark frames in response to external signals passed on some global/bigger than single run start/end scope (ex ever X minutes, every N frames) or on change of  (a sub-set?) of the configuration.

Having 'last dark frame' be state that the detector holds would also make the work flow where the user explicitly takes dark frames 'when they feel like it' possible.

I think we should have a `take_dark_frame` method, but not neccessarily expose it through the BS interface.